src/databricks/labs/ucx/workspace_access/tacl.py

```diff
@@ -21,12 +22,34 @@ def __init__(self, grants_crawler: GrantsCrawler, sql_backend: SqlBackend):
         self._sql_backend = sql_backend
 
     def get_crawler_tasks(self) -> Iterator[Callable[..., Permissions | None]]:
-        def inner(grant: Grant) -> Permissions:
-            object_type, object_key = grant.this_type_and_key()
-            return Permissions(object_type=object_type, object_id=object_key, raw=json.dumps(dataclasses.asdict(grant)))
-
+        # TableAcl grant/revoke operations are not atomic. When granting the permissions,
+        # the service would first get all existing permissions, append with the new permissions,
+        # and set the full list in the database. If there are concurrent grant requests,
+        # both requests might succeed and emit the audit logs, but what actually happens could be that
+        # the new permission list from one request overrides the other one, causing permissions loss.
+        # More info here: https://databricks.atlassian.net/browse/ES-908737
+        #
+        # Below optimization mitigates the issue by folding all action types (grants)
+        # for the same principal, object_id and object_type into one grant with comma separated list of action types.
+        #
+        # For example, the following table grants:
+        # * GRANT SELECT ON TABLE hive_metastore.db_a.table_a TO group_a
+        # * GRANT MODIFY ON TABLE hive_metastore.db_a.table_a TO group_a
+        # will be folded and executed in one statement/transaction:
+        # * GRANT SELECT, MODIFY ON TABLE hive_metastore.db_a.table_a TO group_a
+        folded = collections.defaultdict(lambda: {"action_type": set()})
         for grant in self._grants_crawler.snapshot():
-            yield functools.partial(inner, grant)
+            key = (grant.principal, grant.this_type_and_key())
+            folded[key]["action_type"].add(grant.action_type)
+            grant_dict = dataclasses.asdict(grant)
+            grant_dict["action_type"] = ", ".join(sorted(folded[key]["action_type"]))
+            folded[key]["grant_folded"] = grant_dict
```

Comment:
Why do we need this?

Reply:
we need to fold the action types and emit the just one grant for each key with the updated action type

Reply:
simplified